File: supCPM/approx_geodesic.m
function D = approx_geodesic(W)
% fast geodesic distance matrix computation based on Nystrom method
%
 W = max(W,W');
 n = size(W,1);
 G = graph(W);
 select = 50;
 randindex = randperm(n);
 test_set = randindex(end-10:end);
 test_column = zeros(length(test_set),n)';
 for i=1:length(test_set)
     [T,dist] = shortestpathtree(G,test_set(i),'all','OutputForm','cell');
     test_column(:,i) = dist';
 end
 for draw = 1:floor(n/select)
     for i = (1+(draw-1)*select) : select*draw
        [~,dist] = shortestpathtree(G,randindex(i),'all','OutputForm','cell');
         sample_column(:,i) = dist';
     end
     Y = nystrom(sample_column.^2,randindex(1:i),test_set).^0.5;
     if norm(Y-test_column)<3e-2*norm(test_column)
         break;
         draw
     end
     %temp_st = find(min([sample_column';sample_column'])> prctile(min([sample_column';sample_column']),0));
   %  kk = randperm(length(temp_st));
   %  set = [set,temp_st(kk(1))];
 end
 D = nystrom(sample_column.^2,randindex(1:i),1:n).^0.5;
 D(W~=0)=W(W~=0);
end

function Y = nystrom(X,set,test_set)
n = length(set);
[U S V] = svd(X(set,:)');
n0 = round(n*1/2);
lambda = S(n0,n0);
S(S<=lambda)=0;
Y = X*pinv(V*S*U')*X(test_set,:)';
Y = max(0,Y);
end
 

File: supCPM/R/approx_geodesic.R
approx.geo <- function(W,G){
 n <- length(igraph::V(G))
 select <- 50
 randind = sample(n)
 test_set = tail(randind,11)
 test_col = t(igraph::shortest.paths(G,test_set))
 sample_col = vector()
 for( draw in 1:floor(n/select) ){
   sample_col <- cbind(sample_col,t(igraph::shortest.paths(G,randind[(1+(draw-1)*select) : (select*draw)])))
   Y <- nystrom(sample_col^2,randind[1:(select*draw)],test_set)^0.5
   if( norm(Y-test_col,'2')<3e-2*norm(test_col,'2') ){
     break
   }
 }

 D = nystrom(sample_col^2,randind[1:(select*draw)],1:n)^0.5
 D[W!=0]=W[W!=0]

 return(D)
}


nystrom <- function(X,set,test_set){
  n <- length(set)
  svd_sys <- svd(t(X[set,]),nu=nrow(t(X[set,])),nv=ncol(t(X[set,])))
  U <- svd_sys$u
  V <- svd_sys$v
  S <- matrix(0,nrow(t(X[set,])),ncol(t(X[set,])))
  diag(S) <- svd_sys$d
  n0 <- round(n*0.5)
  lambda <- S[n0,n0]
  S[S<=lambda] <- 0
  Y = X %*% pracma::pinv(V%*%S%*%t(U)) %*% t(X[test_set,])
  Y[Y<0] <- 0
  return(Y)
}
